core/network/stream_manager.hpp

```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_STREAM_MANAGER_HPP
+#define KAGOME_STREAM_MANAGER_HPP
+
+#include <functional>
+#include <memory>
+
+#include <outcome/outcome.hpp>
+
+namespace kagome::network {
+  /**
+   * Manages streams, allowing the client to reuse them
+   * @tparam PeerIdentifier - unique identifier of the peer the stream is opened
+   * to
+   * @tparam ProtocolT - protocol, over which the stream is opened
+   * @tparam Stream - type of the stream
+   */
+  template <typename PeerIdentifier, typename ProtocolT, typename Stream>
+  struct StreamManager {
```

Comment:
I don't think we need to hide libp2p that much using templates. It is fine to expose Stream, PeerId and Protocol types. 
If for some reason we will decide to replace libp2p with something else, I hardly believe it will be possible to use this abstraction to use it with something different from libp2p